extensions/hotreload/hotreload.go

```diff
@@ -0,0 +1,104 @@
+package hotreload
+
+import (
+    "fmt"
+    "html/template"
+    "log"
+    "sync"
+
+    _ "embed"
+
+    . "github.com/emad-elsaid/xlog"
+    "github.com/gorilla/websocket"
+)
+
+var (
+    upgrader     = websocket.Upgrader{ReadBufferSize: 1024, WriteBufferSize: 1024}
+    clients      = make(map[*websocket.Conn]bool)
+    clientsMutex sync.Mutex
+)
+
+func init() {
+    if !READONLY {
+        Listen(Changed, NotifyPageChange)
+        Get(`/ws`, handleWebSocket)
+        RegisterWidget(AFTER_VIEW_WIDGET, 0, clientWidget)
+    }
+}
+
+
+func NotifyPageChange(p Page) error {
+    if !p.Exists() {
+        return nil
+    }
+
+    message := map[string]string{"url": fmt.Sprintf("/%s", p.Name())}
+    log.Printf("Page %s changed from notifiy", p.Name())
+
+    clientsMutex.Lock()
+    defer clientsMutex.Unlock()
+
+    for client := range clients {
+        err := client.WriteJSON(message)
+        if err != nil {
+            log.Printf("Error sending message to client: %v\n", err)
+            client.Close()
+            delete(clients, client)
+        }
+    }
+    return nil
+}
+
+func nop(w Response, r Request) {
+}
+func handleWebSocket(w Response, r Request) Output {
+    conn, err := upgrader.Upgrade(w, r, nil)
+    if err != nil {
+        log.Println("Upgrade error:", err)
+        return BadRequest(err.Error())
+    }
+
+    // keep connection open
+    go func() {
+        defer func() {
+            clientsMutex.Lock()
+            delete(clients, conn)
+            clientsMutex.Unlock()
+            conn.Close()
+        }()
+
+        for {
+            mt, _, err := conn.ReadMessage()
+            if err != nil || mt ==  websocket.CloseMessage {
+                break
+            }
+        }
+    }()
+
+    clientsMutex.Lock()
+    clients[conn] = true
+    clientsMutex.Unlock()
+    log.Printf("Connection arrived. curr cons: %d", len(clients))
+    return nop
+}
+
+// TODO use same HOST and PORT than server
+const clientScript = `
+    <script>
+    (() => {
+        const socketUrl = 'ws://localhost:3000/ws';
+        let socket = new WebSocket(socketUrl);
+        socket.addEventListener('message', (evt) => {
+            if (evt.data.url != window.location.href) {
```

Comment:
evt.data at this point is a string, so `.url` is undefined. so I think you'll need to move line 93 before 92 to parse first then use `data.url`

The other point is `window.location.href` returns the  whole URL `http://localhost:3000/pagename` not `/pagename` and the `url` you passed in the event is just `/pagename` so it'll never match

but also what if it matches? shouldn't we reload the page to show the changes? 

So I think this section should be without this if statement, and directly parse and set the href. it should navigate if the page is different and reload if not. and I think we can use `.path` instead of `href`

Reply:
oh yes. that's true. and yes, better without the if statement. I'm already using it as it is, and it reloads every time, as expected not expected